### PR TITLE
Fixing a small bug in ribo_frac. 

### PR DESCRIPTION
**Pull request by simonleandergrimm**
ribo_frac currently breaks when there are no ribosomal reads, as we try to create an average with weights that sum to zero. I've created a simple fix by creating a `try except` loop that sets weighted_rrna_fraction to 0 in such cases.

See error that was previously thrown:
```
[ec2-user@ip-172-31-32-232 mgs-pipeline]$ ./run.py --skip-stages riboreads --bioproject=PRJNA395239
Skipping SRR5853126
Traceback (most recent call last):
  File "/home/ec2-user/code/mgs-pipeline/./run.py", line 1551, in <module>
    start()
  File "/home/ec2-user/code/mgs-pipeline/./run.py", line 1547, in start
    STAGE_FNS[stage](args)
  File "/home/ec2-user/code/mgs-pipeline/./run.py", line 460, in ribofrac
    weighted_rrna_fraction = np.average(
  File "/home/ec2-user/.local/lib/python3.9/site-packages/numpy/lib/function_base.py", line 550, in average
    raise ZeroDivisionError(
ZeroDivisionError: Weights sum to zero, can't be normalized
```

### CODE REVIEW

File: run.py [hunk 1]
```diff
@@ -454,12 +454,15 @@ def calculate_average_read_length(file_path):
...
         # Use the total number of reads for each input as weights
         weights = list(total_reads_dict.values())
-        weighted_rrna_fraction = np.average(
-            fractions_rrna_in_subset, weights=weights
-        )
+        try:
+            weighted_rrna_fraction = np.average(
+                fractions_rrna_in_subset, weights=weights
+            )
+        except:
```
Comment (jeffkaufman) [inline run.py]: Don't use bare `except`: instead list the specific exception you want to catch:

```
except ZeroDivisionError:
   ...
```

Comment (simonleandergrimm) [reply run.py]: fixed

File: run.py [hunk 2]
```diff
@@ -454,12 +454,15 @@ def calculate_average_read_length(file_path):
...
-        weighted_rrna_fraction = np.average(
-            fractions_rrna_in_subset, weights=weights
-        )
+        try:
+            weighted_rrna_fraction = np.average(
+                fractions_rrna_in_subset, weights=weights
+            )
+        except ZeroDivisionError:
+            weighted_rrna_fraction = 0
+            print("Sample doesn't contain ribosomal reads.")
```
Comment (jeffkaufman) [inline run.py]: Is this actually what a ZeroDivisionError means here? I'd expect it to happen when there were zero reads, not zero ribosomal reads.

Comment (simonleandergrimm) [reply run.py]: @lennijusten can you say more? I'm not super familiar with this code.

File: run.py
Comment (lennijusten) [inline run.py:1]: I'd actually expect the `file_integrity_check(filename)` func to catch files with no reads so I'm surprised this error came up. Note lines 320-328

```
with gzip.open(filename, "rt") as f:
    try:
        # Check for the first read using the iterator
        next(FastqGeneralIterator(f))
    except StopIteration:
        warnings.warn(f"File {filename} contains no reads!")
        return False
```

File: run.py (hunk 2 above)
Comment (lennijusten) [reply run.py]: `weights` is calculated using the total reads in each subset where subsets are subsets of the different adapter trimming output files. I would expect weights to sum to zero if there were zero reads. 

So @jeffkaufman is right. If we get the `ZeroDivisionError` it means we have zero total reads and hence also zero rRNA reads.

Comment (simonleandergrimm) [reply run.py]: Ok cool, I'll revisit this (might also check in with you in person @lennijusten)

**Comment by simonleandergrimm** [comment]
I dug into this some more. 

Current takeaway:

`ribo_frac` fails because there are no files for `SRR5853126_1.*` in `s3://nao-mgs/PRJNA395239/cleaned/`. This isn't caught because though `ribofrac` checks if a file with `SRR5853126` exists in `cleaned`, it doesn't raise an error if it doesn't.
```
for sample in get_samples(args):
        # Check for name of output file
        sample_output_file = sample + ".ribofrac.txt"
        if sample_output_file in existing_outputs:
            continue

        # Track for error handling
        total_files_in_sample = 0
        empty_files_in_sample = 0

        total_reads_dict = {}
        subset_reads_dict = {}
        rrna_reads_dict = {}
        for potential_input in available_inputs:
            if not potential_input.startswith(sample):
                continue
            if ".settings" in potential_input:
                continue
            if "discarded" in potential_input:
                continue
```
I.e., when I debugged this, `SRR5853126` never passed beyond:

```
for potential_input in available_inputs:
    if not potential_input.startswith(sample):
        continue
    if ".settings" in potential_input:
        continue
    if "discarded" in potential_input:
        continue
```

But the code continues to execute after the `for potential_input in available_inputs` loop:
```
for potential_input in available_inputs:
   [a loop that doesn't run on SRR5853126...]

# Calculate the weighted average fraction of rRNA reads across all inputs in sample using numpy
# Extract the fractions of rRNA reads for each input
fractions_rrna_in_subset = [
    rrna_reads_dict[input_filename] / subset_reads_dict[input_filename]
    for input_filename in total_reads_dict
]

# Use the total number of reads for each input as weights
weights = list(total_reads_dict.values())
weighted_rrna_fraction = np.average(
    fractions_rrna_in_subset, weights=weights
)
fraction_rrna = round(weighted_rrna_fraction, 4)

print(
    f"Estimated fraction of rRNA reads in {sample} = {round(fraction_rrna*100, 2)}%"
)
```


There are no files for `SRR5853126` in `/cleaned` because it is skipped in `adapter_removal` due to `s3://nao-mgs/PRJNA395239/raw/SRR5853126_1.fastq.gz` being 0B. See relevant code in `adapter_removal`:
```
def adapter_removal(args, dirname, trim_quality, collapse):
    adapter_dir = work_fname("bioprojects", args.bioproject, "adapters")
    try:
        os.mkdir(adapter_dir)
    except FileExistsError:
        pass

    available_inputs = get_files(args, "raw")
    existing_outputs = get_files(args, "cleaned", min_size=100)

    for sample in get_samples(args):
        raw1 = "%s_1.fastq.gz" % sample
        raw2 = "%s_2.fastq.gz" % sample

        if raw1 not in available_inputs or raw2 not in available_inputs:
            print("Skipping %s" % sample)
            continue
```

@jeffkaufman Probably we or you want to fix this on Monday, I'm not sure where we want to fix the code to stop this from happening.

FWIW, the raw FASTQ of SRR5853126 looks like this:
```
(venv) simongrimm@Simons-MacBook-Pro-3 ~ % gunzip Downloads/SRR5853126.fastq.gz 
(venv) simongrimm@Simons-MacBook-Pro-3 ~ % head -50 Downloads/SRR5853126.fastq
@SRR5853126.1.1 1 length=251
AGACGGGCTAGGTCGGGGTAATATTCAGGCTCCGGGCGGTAGGCCGTGAAGCTTCCGCATCAAGTTTTTGGCTGTATCGAACGGCCCGCGGTCCAAAGCGGCGTCAATGTCGCCGCGTCCGTACTCTCGCGCCGTGGAGCCGATCTTATTGACCCTTTCCCAGATCGGAAGAGCACACGTCTGAACTCCAGTCACATTCAGAAATATCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAATTAAACTACACA
+SRR5853126.1.1 1 length=251
???????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????
@SRR5853126.1.2 1 length=251
GGGAAAGGGTCAATAAGATCGGCTCCACGGCGCGAGAGTACGGACGCGGCGACATTGACGCCGCTTTGGACCGCGGGCCGTTCGATACAGCCAAAAACTTGATGCGGAAGCTTCACGGCCTACCGCCCGGAGCCTGAATATTACCCCGACCTAGCCCGTCTAGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGGATAGGGTGGATAGCTCGATGGCGGCCGTGGGGTTAAAGAACGAAGAGGAGTCAA
+SRR5853126.1.2 1 length=251
???????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????
@SRR5853126.2.1 2 length=251
```


@lennijusten flagging your for awareness

**Comment by lennijusten** [comment]
Thanks Simon. We can discuss. I wrote the code so I'm happy to try and submit a fix.

**Comment by simonleandergrimm** [comment]
@jeffkaufman The PR is ready for review again. I fixed the bug by creating a flag that tracks if any file for a sample is present in `/cleaned`.

File: run.py [hunk 3]
```diff
@@ -447,6 +448,10 @@ def calculate_average_read_length(file_path):
                     )
                 )
                 rrna_reads_dict[inputs[0]] = subset_reads - non_rrna_count
+        if not file_present:
```
Comment (jeffkaufman) [inline run.py]: How about `if not total_files_in_sample`, which means you don't need `file_present` at all?

**Review by jeffkaufman — APPROVED** [review]

File: run.py (hunk 3 above)
Comment (simonleandergrimm) [reply run.py]: Yep, that's nicer. Fixed!

**Pull request merged**